lib/freebl/algcmac.c

```diff
@@ -0,0 +1,327 @@
+            if (cmac_UpdateState(ctx) != SECSuccess) {
+                return SECFailure;
+            }
+
+            ctx->partialIndex = 0;
+        }
+
+        unsigned int copy_len = data_len - data_index;
+        if (copy_len > (ctx->blockSize - ctx->partialIndex)) {
+            copy_len = ctx->blockSize - ctx->partialIndex;
+        }
+
+        PORT_Memcpy(ctx->partialBlock + ctx->partialIndex, data + data_index, copy_len);
+        data_index += copy_len;
+        ctx->partialIndex += copy_len;
+    }
+
+    return SECSuccess;
+}
+
+SECStatus CMAC_Finish(CMACContext *ctx, unsigned char *result,
+                      unsigned int *result_len,
+                      unsigned int max_result_len) {
+    if (ctx == NULL || result == NULL || max_result_len == 0) {
+        PORT_SetError(SEC_ERROR_INVALID_ARGS);
+        return SECFailure;
+    }
+
+    if (max_result_len > ctx->blockSize) {
+        // This is a weird situation. The PKCS #11 soft token code passes
+        // sizeof(result) here, which is hard-coded as SFTK_MAX_MAC_LENGTH.
+        // This later gets truncated to min(SFTK_MAX_MAC_LENGTH, requested).
+        max_result_len = ctx->blockSize;
+    }
+
+    // Step 4: If M_n* is a complete block, M_n = K1 ^ M_n*. Else,
+    // M_n = K2 ^ (M_n* || 10^j).
+    if (ctx->partialIndex == ctx->blockSize) {
+        // XOR in K1.
+        for (unsigned int index = 0; index < ctx->blockSize; index++) {
+            ctx->partialBlock[index] ^= ctx->k1[index];
+        }
+
+        // Encrypt the block.
+        if (cmac_UpdateState(ctx) != SECSuccess) {
+            return SECFailure;
+        }
+    } else {
+        // Use 10* padding on the partial block.
+        ctx->partialBlock[ctx->partialIndex++] = 0x80;
+        PORT_Memset(ctx->partialBlock + ctx->partialIndex, 0,
+                    ctx->blockSize - ctx->partialIndex);
+        ctx->partialIndex = ctx->blockSize;
+
+        // XOR in K2.
+        for (unsigned int index = 0; index < ctx->blockSize; index++) {
+            ctx->partialBlock[index] ^= ctx->k2[index];
+        }
+
+        // Encrypt the block.
+        if (cmac_UpdateState(ctx) != SECSuccess) {
+            return SECFailure;
+        }
+    }
+
+    // Step 7 & 8: T = MSB_tlen(C_n); return T.
+    PORT_Memcpy(result, ctx->lastBlock, max_result_len);
+    if (result_len != NULL) {
+        *result_len = max_result_len;
+    }
+    return SECSuccess;
+}
+
+void CMAC_Destroy(CMACContext *ctx, PRBool free_it) {
+    if (ctx == NULL) {
+        return;
+    }
+
+    if (ctx->cipher.raw != NULL) {
+        if (ctx->cipherType == CMAC_AES) {
+            AES_DestroyContext(ctx->cipher.aes, PR_TRUE);
+        }
+    }
+
+    PORT_ZFree(ctx->k1, ctx->blockSize);
+    PORT_ZFree(ctx->k2, ctx->blockSize);
+    PORT_ZFree(ctx->partialBlock, ctx->blockSize);
+    PORT_ZFree(ctx->lastBlock, ctx->blockSize);
+
+    if (ctx->wasAllocated && free_it == PR_TRUE) {
+        PORT_ZFree(ctx, sizeof(*ctx));
+    }
```

Comment:
May still need to zero ctx to clear out any CPSs.

Reply:
Ah right, thanks! I'll put that ahead of the `PORT_ZFree` call and switch it to a `PORT_Free`.

Reply:
Now that I think about it, there's no CPS leakage here. All CPS is stored in:

 - `cipher` (a pointer, which it is up to the block cipher context to clear and free). Memsetting this won't do anything but clear a pointer to what should now be a zero'd segment of memory (if `AES_DestroyContext` does its job that is. If it doesn't memset won't really help much).
 - `k1` (a pointer, which `PORT_ZFree` should clear the contents of).
 - `k2` (a pointer, which `PORT_ZFree` should clear the contents of).
 - `partialBlock` (a pointer, which `PORT_ZFree` should clear the contents of)
 - `lastBlock` (a pointer, which `PORT_ZFree` should clear the contents of) -- this is the output of the MAC at this point, of questionable sensitivity. 

So a `PORT_Memset` over the contents of the structure at best prevent us from leaking pointers to cleared structures.

Thoughts?

Reply:
(If with `left_shift`, we decide to allocate `k1` / `k2` directly in `CMACContextStr`, we would then have cause to `PORT_Memset`).

Reply:
OK, I was thinking ctx was ctx->cipher. The current code is good. Of course if you implement my suggestion making k1, k2, partialBlock and lastBlock arrays, then you'd need the PORT_Memset again. If you don't, just include your comment above as a comment in the code so FIPS reviewers have an easier time.